CONTRIBUTING.md

```diff
@@ -0,0 +1,173 @@
+  with additional information.
+* Submitting translation improvements via Crowdin (see below).
+* Contributing code changes with pull requests (PRs).
+
+## Contributing translations
+
+If you would like to contribute translations, you can submit improvements to
+[our project on Crowdin](https://crowdin.com/project/xournalpp). The Crowdin
+translations are merged back into the main code periodically, after which they
+become available available in the nightly builds.
+
+## Contributing code changes with pull requests (PRs)
+
+Xournal++ development primarily occurs on [GitHub at the xournalpp/xournalpp
+repository](https://github.com/xournalpp/xournalpp). As a contributor, you
+probably have a particular bug or feature that you are interested in working on.
+Before you start, you should first look in the [issue
+tracker](https://github.com/xournalpp/xournalpp/issues) to see if anyone has
+reported your bug or has a similar idea for a feature. If not, you should first
+submit a new issue detailing what you are about to do. This will allow you to
+get feedback and ensure that you do not end up duplicating work. You can obtain
+additional help by contacting community members through one of our [official
+communication channels](https://xournalpp.github.io/community/help/).
+
+The process for contributing code changes works as follows:
+
+1. Fork the main xournalpp repository and create commits on your own feature
+   branch.
+2. Submit a pull request (PR).
+3. Wait for maintainers to review the PR and address the relevant feedback.
+4. After receiving maintainer approval, the PR is merged after a short grace
+   period.
+
+### Creating a fork
+
+To make a code change, you'll need to first fork the repository on GitHub, if
+you haven't already. To do so, click the "Fork" button in the top right corner
+of the repository's web page.
+
+We assume that you are already familiar with Git, which we use to track changes.
+An excellent resource for Git is the _Pro Git_ book by Scott Chacon and Ben
+Straub, which is available online for free
+[here](https://git-scm.com/book/en/v2).
+
+You should try to first clone your fork and then compile it manually. The
+instructions for compiling Xournal++ on your operating system can be found at
+[`LinuxBuild.md`](readme/LinuxBuild.md), [`MacBuild.md`](readme/MacBuild.md),
+and [`WindowsBuild.md`](readme/WindowsBuild.md). Once you are set up, you will
+be ready to make code changes.
+
+### Code conventions and guidelines
+
+Here are some of the _guidelines_ that we follow when working on the Xournal++
+codebase. We emphasize _guidelines_ because these are ideal points that we'd
+like to see in contributions, but we understand that perfectly following all
+points is very difficult in practice.
+
+General contribution guidelines:
+
+* **The code should be developed on a Linux distribution (of your choosing)**.
+  The project is set up to _build_ on MacOS or Windows but is not ideal for
+  development there.
```

Comment:
Is this 'by design' or just because so far nobody cared or had the time to work out a decent development setup – or is there an inherent reason I am missing?

In the second case something can be done about this and I'd explicitly state something like

```suggestion
* **So far Linux is the most convenient platform for code development.** The
  documented setup on MacOS and Windows is not ideal, but if you have an idea,
  you're welcome to share it in a PR!
```

In the first case ... well, I'll have to dump my work on Windows console and crash handlers then (yes, this comment is completely self-serving (: ).

But jokes aside, my (admittedly unqualified) opinion is that, for a multi platform app, it helps if development happens on all target platforms.

Reply:
Thanks for the suggestion. The answer to your question is simply:

>  so far nobody cared or had the time to work out a decent development setup

The Windows setup relies on Msys2, which assumes familiarity with Linux. But the majority of people familiar with Linux are likely using Xournal++ on Linux rather than on Windows with Msys2.

> for a multi platform app, it helps if development happens on all target platforms.

I agree with you, but all the core maintainers are each using some Linux distribution as a daily driver. Due to limitations on time, the extent to which we "develop" on Windows and MacOS is limited to fixing platform-specific bugs and testing the packaging.

I think setting up development for Windows is actually not too bad: steps are just install dependencies & run build scripts to the developer. IDE configuration is undocumented, though.

On the other hand, trying to develop for MacOS means having to go through the complexity of setting up jhbuild and compiling GTK from source.

Reply:
Thanks for your answer!

> Due to limitations on time, the extent to which we "develop" on Windows and MacOS is limited ...

This is understandable of course, but makes community input all the more important, doesn't it? That's why I suggest to use a more encouraging wording. I just think "You should use Linux" might put off potential contributors, which would be a pity because, as you said:

> setting up development for Windows is actually not too bad

... except for horrible build times with GCC, but this is fixable, see below.

> The Windows setup relies on Msys2, which assumes familiarity with Linux. But the majority of people familiar with Linux are likely using Xournal++ on Linux rather than on Windows with Msys2.

I wouldn't be so sure here. I believe I'm not the only one using Xournal++ on Windows as a simpler/more transparent/just better fitting alternative to OneNote for taking notes on lectures primarily; thus I'd rather not fiddle around with a Linux install on my Surface Book 2 just for development. Installing MSYS2 etc. is much simpler compared to that.
(That's not to say Linux is bad, I have a dual boot machine set up myself, just not the machine I use Xournal++ on. And coding on one machine, then packaging and transferring the whole build to the other to test it will get tedious.)

> IDE configuration is undocumented, though.

I can try to fix this once I've got some more time and figured out the last details.

<details><summary>(OT) Sneak peek: developing on Windows – will submit/update PRs when ready...</summary>
I'm using VS Code with C++, clangd, CMake extensions and build with Clang from the MSYS2/clang64 environment, which builds much faster than GCC, but needed some adjustments with the packaging script. Three compiler/linker flags later Clang emits PDB symbols, so now Xournal++ can be debugged with the apparently so popular VS debugger (compared to gdb it's really ... fast).<br>
Along the way this enabled rich stacktraces using the Win32 dbghelp API, so with crash handlers  implemented (which I did for #3746 initially), Windows users will also be able to file meaningful bug reports (#3802 being a counterexaple of 'meaningful').<br>
<b>TL;DR:</b> development on Windows does have some potential.</details>

Reply:
Thanks for the feedback, I incorporated it into the updated bullet point. Let's move the discussion into a separate issue, as I think it would be more visible to other people developing on Windows.